### PR TITLE
Websocket module

### LINKED ISSUE #5806: WebSockets for Godot
WebSockets are really simple to use, esp. when creating multiplayer games. I am suggesting there be a WebSocket node in Godot 3. That way one can easily create a client & hook it up to an existing server effortlessly.

I don't know if WebSockets are already supported in the current networking model, but so far I haven't seen mention of it anywhere & I wouldn't know how to implement it using the HTTP nodes. Sorry if this issue is a result of my ignorance.

### PR DESCRIPTION
**Pull request by Faless**
Webassembly support is client-only for obvious reasons.
Other platforms support both client and server using libwebsockets.

Libwebsockets is released under LGPLv2.1 + static linking exception (and others, like not having to distribute the license itself, see the `LICENSE.txt` for details).

The module itself can easily work with other libraries, since it's done with custom implementations of generic `WebSocketClient` `WebSocketServer` and `WebSocketPeer` classes to allow for webassembly native implementation.

The module also supports the **Multiplayer API**, but **beware** that websocket uses **TCP**, and it's always reliable. This means that **latency will grow indefinitely if you send too many RPCs**.
This is fine for most non-realtime games, e.g. turn based, browser games, etc. and it's a first step toward having both web and server implementations done with godot, same scene, seamlessly.

The next step, when time allows, would clearly be WebRTC.

EDIT:
Closes #5806.
Brings #14627 to a better state.
Partly unrelated but #9624 should be closed as HTTPClient works in 3.0 for webassembly export.

### CODE REVIEW

**Comment by Faless** [comment]
I can't explain the failed build on X11 platform. It's my main platform, and it builds fine here with both builtin and distro's openssl

**Comment by Faless** [comment]
In any case SSL with lws is not yet implemented properly, so I might just disable it for now.

**Comment by akien-mga** [comment]
Also builds fine for me on X11. Might be related to the OpenSSL version on Travis/Ubuntu 14.04?

**Comment by Faless** [comment]
> Might be related to the OpenSSL version on Travis/Ubuntu 14.04?

You are totally right, apparently LWS requires OpenSSL >= 1.0.2 and ubuntu 14.04 seems to still have OpenSSL 1.0.1f . I for once agree with you @akien-mga that ubuntu packages sucks.
See https://github.com/warmcat/libwebsockets/issues/666#issuecomment-259746135 and https://github.com/warmcat/libwebsockets/commit/e7bf0aa1dcaa73171ef300d36ae0b6c3cc5140da

I see that function is protected by an ifdef block, I can try to disable SSL hostname verification and maybe try to implement it in a compatible way later.

**Comment by Faless** [comment]
Okay, it builds now, I'm pretty positive the client no longer does SSL hostname verification though.

We could fix it in the lws callback. I'm not sure if I'll have the time before January, but I'll try.

**Comment by akien-mga** [comment]
Otherwise we can just build on travis against the builtin openssl which is 1.0.2 (might increase build time a bit, but with the cache it shouldn't be too bad).

**Comment by Faless** [comment]
> Otherwise we can just build on travis against the builtin openssl which is 1.0.2 (might increase build time a bit, but with the cache it shouldn't be too bad).

I guess official export templates for linux are shipping with builtin openssl.
In that case I'd say let's just bump openssl min version requirements :)

**Comment by akien-mga** [comment]
> I guess official export templates for linux are shipping with builtin openssl.

Yep: https://github.com/GodotBuilder/godot-builds/blob/master/.travis.yml#L26

> In that case I'd say let's just bump openssl min version requirements :)

Go ahead. You can remove your last commit then and tweak `.travis.yml` so that it builds against the builtin openssl.

File: .travis.yml [hunk 1]
```diff
@@ -8,6 +8,7 @@ env:
   global:
     - SCONS_CACHE=$HOME/.scons_cache
     - SCONS_CACHE_LIMIT=1024
+    - OPTIONS="verbose=yes progress=no openmp=no"
```
Comment (Faless) [inline .travis.yml]: @akien-mga  you said tweak, I tweaked :wink:

**Comment by akien-mga** [comment]
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

**Comment by akien-mga** [comment]
It seems to fail building on X11 and Windows.

**Comment by hpvb** [comment]
also the mbed_tls stuff should probably be removed. It seems like a bad idea to have two separate ssl implementations in the tree.

**Comment by reduz** [comment]
Actually if anyone wants to replace openssl with mbedtls in Godot, that
could be good.. right?

**Comment by hpvb** [comment]
@reduz I guess that'll depend on whether or not curl can work with mbedtls.

**Comment by Faless** [comment]
> It seems to fail building on X11 and Windows.

Yes, my bad, I forgot to update it to comply the last minute change to the NetworkedMultiplayerPeer API (`peer.transfer_mode` property).
Should compile again now (does for me on X11)

**Comment by akien-mga** [comment]
I'd also support dropping openssl in favour of a more lightweight alternative.

At least in raw number of files, mbedtls seems quite smaller than openssl: https://github.com/ARMmbed/mbedtls/tree/development/library https://github.com/ARMmbed/mbedtls/tree/development/include/mbedtls

**Comment by Faless** [comment]
> also the mbed_tls stuff should probably be removed. It seems like a bad idea to have two separate ssl implementations in the tree.

@hpvb  the `mbed_tls` code does not get included by `modules/websocket/SCsub` so it does not (should not?!) be compiled at all (it doesn't in my tests).

It's left there because I prefer to not fiddle around with library files and allow to simply copy/paste them from the library source when a new version is released. It's true that it makes the Godot source tree slightly bigger but the resulting binary is not affected at all

**Comment by hpvb** [comment]
@Faless we've previously always deleted files that we don't build from thirdparty/ (look at what was done to thekla_unwrap for instance).

**Comment by akien-mga** [comment]
Indeed, we always copy only what we need from thirdparty sources, which is why https://github.com/godotengine/godot/blob/master/thirdparty/README.md has sections describing for all libraries what files should be extracted from the upstream tarball or repo.

**Comment by Faless** [comment]
> we've previously always deleted files that we don't build from thirdparty/ (look at what was done to thekla_unwrap for instance).

I see, well, I'll check what files I can remove then.

**Comment by akien-mga** [comment]
Could you squash that last commit into the first one? This way we'll never have added this code to the repo ;)

**Comment by Faless** [comment]
> Could you squash that last commit into the first one? This way we'll never have added this code to the repo ;)

Sure, I'll wait for the tests to finish so I'm sure it compiles everywhere (did some test here, but only cross-compiling from Linux)

**Comment by Faless** [comment]
Sqaushed commits, updated to latest Emscripten API

**Comment by LikeLakers2** [comment]
Apologies if I'm missing something, but I'm looking at the files under `modules/websocket/`, and I'm noticing several instances under the `lws_*.cpp` and `emws_*.cpp` files where there's a `get_connected_port` function. At a guess, this is probably a function to get the port that's being used for the connection. So assuming my guess is right, why are all of them set to return a static `1025` rather than the port variable? It just seems odd, like it was being used for debug and wasn't removed for production.

**Comment by Faless** [comment]
If I recall correctly the interface requires it, but it was never
implemented.

If you need it, open a new issue on GitHub.

Otherwise, if the interface does not require it, we might just remove it.

On Feb 7, 2018 06:53, "LikeLakers2" <notifications@github.com> wrote:

Apologies if I'm missing something, but I'm looking at the files under
modules/websocket/, and I'm noticing several instances under the lws_*.cpp
and emws_*.cpp files where there's a get_connected_port function. At a
guess, this is probably a function to get the port that's being used for
the connection. So assuming my guess is right, why are all of them set to
return a static 1025 rather than the port variable? It just seems odd, like
it was being used for debug and wasn't removed for production.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/godotengine/godot/pull/14888#issuecomment-363665266>,
or mute the thread

**Comment by matty** [comment]
Would just like to add that I have built Godot from _master_ on Win 10 and I am using the WebSocket implementation and it's working well so far.

**Comment by Faless** [comment]
@endel when using `gd_mp_api=false` (default) both WebSocketClient and WebSocketServer simply acts as a standard-conformant implementation of RFC6455, so they will be compatible to any other RFC6455-conformant software.
If you enable `gd_mp_api=true` (i.e. you use it for high level multiplayer API), the internal Godot protocol for peer exchange and rpc/rset relaying is used on top of the standard WebSocket protocol (so it will be compatible, but data will be always transferred as binary, and will contain additional protocol data).

**Comment by joaopedrosgs** [comment]
there is any documentation about this?

**Comment by Faless** [comment]
@joaopedrosgs well, there's the official doc for the WebSocketClient class: http://docs.godotengine.org/en/latest/classes/class_websocketclient.html#description

**Comment by vini-guerrero** [comment]
Is there any tutorial on connectin this with a simple nodejs server? ( is that possible? )

**Pull request merged**